sycl/include/CL/sycl/ONEAPI/intel_matrix/matrix-amx.hpp

```diff
@@ -0,0 +1,438 @@
+
+// This handles cases where T1 is int16(bfloat16), T2 is float.
+inline __SYCL_ALWAYS_INLINE static void
+submatrix_mad(detail::submatrix<unsigned short> &sub_ma,
+              detail::submatrix<unsigned short> &sub_mb,
+              detail::submatrix<float> &sub_mc) {
+  sub_mc.tile =
+      matrix::tdpbf16ps_internal(sub_mc.rows, sub_mc.cols, sub_ma.cols,
+                                 sub_mc.tile, sub_ma.tile, sub_mb.tile);
+}
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols>
+inline __SYCL_ALWAYS_INLINE static
+    typename std::enable_if<(NumRows > matrix::tile_size) ||
+                                (NumCols * sizeof(T) / 4 > matrix::tile_size),
+                            void>::type
+    submatrix_store(detail::submatrix<T> &sub_m,
+                    matrix::joint_matrix<Group, T, NumRows, NumCols> &jm,
+                    uint32_t row, uint32_t col, size_t stride,
+                    matrix::matrix_layout layout, bool shouldreload) {
+  uint32_t offset = (row * stride + col);
+  T *ptr = (T *)jm.raw_storage;
+  ptr += offset;
+  stride *= sizeof(T);
+  matrix::tilestored64_internal(sub_m.rows, sub_m.cols, (char *)ptr, stride,
+                                sub_m.tile);
+}
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols>
+inline __SYCL_ALWAYS_INLINE static
+    typename std::enable_if<(NumRows <= matrix::tile_size) &&
+                                (NumCols * sizeof(T) / 4 <= matrix::tile_size),
+                            void>::type
+    submatrix_store(detail::submatrix<T> &sub_m,
+                    matrix::joint_matrix<Group, T, NumRows, NumCols> &jm,
+                    uint32_t row, uint32_t col, size_t stride,
+                    matrix::matrix_layout layout, bool shouldreload) {
+  if (shouldreload) {
+    int8_t NewjmC[matrix::tile_size * matrix::tile_size * 4];
+    matrix::tilestored64_internal(matrix::tile_size, matrix::tile_size * 4,
+                                  (char *)NewjmC, matrix::tile_size * 4,
+                                  sub_m.tile);
+    jm.tile = matrix::tileloadd64_internal(
+        NumRows, NumCols * sizeof(T), (char *)NewjmC, matrix::tile_size * 4);
+    return;
+  }
+  jm.tile = sub_m.tile;
+}
+
+} // namespace detail
+
+namespace matrix {
+
+// This handles cases where matrix can't be accommodated by a tile
+template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          matrix_layout Layout, access::address_space Space>
+inline __SYCL_ALWAYS_INLINE typename std::enable_if<
+    (NumRows > tile_size) || (NumCols * sizeof(T) / 4 > tile_size), void>::type
+joint_matrix_load(Group sg,
+                  joint_matrix<Group, T, NumRows, NumCols, Layout> &jm,
+                  multi_ptr<T, Space> src, size_t stride,
+                  matrix_layout layout) {
+  T *mem = src.get();
+  // memcpy from mem to jm.raw_storage
+  for (int i = 0; i < NumRows; ++i) {
+    char *srcptr = (char *)mem + i * stride * sizeof(T);
```

Comment:
Use C++ casts instead?

Reply:
yeah, I will change it. BTW, is there any advantage of reinterpret_cast, compared with c-style cast? Is that because it is more eye-catching?